File: benchmarks/tree_traversal/cpp/mem.cpp
// Measures peak heap allocation of compact_tree traversal at 100_000 tips.
// Prints: {"peak_bytes": <N>}

#include "../../common/cpp/mem_track.hpp"

#include <cstdio>
#include <fstream>
#include <stdexcept>
#include <string>

#include "compact_tree.h"

static std::string load_file(const std::string& path)
{
    std::ifstream ifs(path, std::ios::binary);
    if (!ifs)
        throw std::runtime_error("Cannot open " + path +
            " - run benchmarks/data/prepare_data.py first");
    std::string s((std::istreambuf_iterator<char>(ifs)), {});
    while (!s.empty() && (s.back() == '\n' || s.back() == '\r'))
        s.pop_back();
    return s;
}

int main()
{
    const auto newick = load_file("../../data/prepared/tree_100000.nwk");

    mem_track::reset();

    compact_tree tree(newick, /*is_fn=*/false);
    double total = 0.0;
    for (auto it = tree.postorder_begin(); it != tree.postorder_end(); ++it)
        total += static_cast<double>(tree.get_edge_length(*it));

    if (total < 0) std::printf("(impossible)\n");

    const std::size_t peak = mem_track::peak_bytes();
    std::printf("{\"peak_bytes\": %zu}\n", peak);
    return 0;
}

File: benchmarks/tree_traversal/rust/examples/mem.rs
// measures peak heap allocation of traverse_newick at n=100_000.
// prints {"peak_bytes": <N>} to stdout.

#[global_allocator]
static ALLOC: dhat::Alloc = dhat::Alloc;

use std::hint::black_box;
use std::path::Path;

fn data_path(file: &str) -> std::path::PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("../../data/prepared")
        .join(file)
}

fn main() {
    let newick = std::fs::read_to_string(data_path("tree_100000.nwk"))
        .expect("tree_100000.nwk not found - run benchmarks/data/prepare_data.py first");

    let _profiler = dhat::Profiler::builder().testing().build();

    black_box(tree_traversal::traverse_newick(black_box(&newick)));

    let stats = dhat::HeapStats::get();
    println!("{{\"peak_bytes\": {}}}", stats.max_bytes);
}
